c/tskit/trees.c

```diff
@@ -2124,6 +2124,530 @@ tsk_treeseq_sample_count_stat(const tsk_treeseq_t *self, tsk_size_t num_sample_s
     return ret;
 }
 
+/***********************************
+ * Two Locus Statistics
+ ***********************************/
+
+static int
+get_allele_samples(const tsk_site_t *site, const tsk_bit_array_t *state,
+    tsk_bit_array_t *out_allele_samples, tsk_size_t *out_num_alleles)
+{
+    int ret = 0;
+    tsk_mutation_t mutation, parent_mut;
+    tsk_size_t mutation_index, allele, alt_allele_length;
+    /* The allele table */
+    tsk_size_t max_alleles = site->mutations_length + 1;
+    const char **alleles = tsk_malloc(max_alleles * sizeof(*alleles));
+    tsk_size_t *allele_lengths = tsk_calloc(max_alleles, sizeof(*allele_lengths));
+    const char *alt_allele;
+    tsk_bit_array_t state_row;
+    tsk_bit_array_t allele_samples_row;
+    tsk_bit_array_t alt_allele_samples_row;
+    tsk_size_t num_alleles = 1;
+
+    if (alleles == NULL || allele_lengths == NULL) {
+        ret = TSK_ERR_NO_MEMORY;
+        goto out;
+    }
+
+    tsk_bug_assert(state != NULL);
+    alleles[0] = site->ancestral_state;
+    allele_lengths[0] = site->ancestral_state_length;
+
+    for (mutation_index = 0; mutation_index < site->mutations_length; mutation_index++) {
+        mutation = site->mutations[mutation_index];
+        /* Compute the allele index for this derived state value. */
+        for (allele = 0; allele < num_alleles; allele++) {
+            if (mutation.derived_state_length == allele_lengths[allele]
+                && tsk_memcmp(
+                       mutation.derived_state, alleles[allele], allele_lengths[allele])
+                       == 0) {
+                break;
```

Comment:
We should be covering this, really. Do we need an example with some extra mutations or something?

Reply:
I added a test with a couple of back-mutations, that covers it